scripts-available/CDB_FederatedServer.sql

```diff
@@ -0,0 +1,395 @@
+            -- TODO: Delete this IF before merging to make sure nobody creates a role
+            -- that is later used automatically by us granting them all permissions in the foreign server
+            -- TODO: This is here to help debugging during development (so failures to destroy objects are allowed)
+            -- TODO
+            IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname = role_name) THEN
+                EXECUTE FORMAT('CREATE ROLE %I NOLOGIN', role_name);
+            END IF;
+            EXECUTE FORMAT('GRANT ALL PRIVILEGES ON DATABASE %I TO %I', current_database(), role_name);
+            EXECUTE FORMAT('GRANT USAGE ON FOREIGN DATA WRAPPER postgres_fdw TO %I', role_name);
+            EXECUTE FORMAT('GRANT USAGE ON FOREIGN DATA WRAPPER postgres_fdw TO %I', role_name);
+            EXECUTE FORMAT('GRANT USAGE ON FOREIGN SERVER %I TO %I', server_internal, role_name);
+            EXECUTE FORMAT('ALTER SERVER %I OWNER TO %I', server_internal, role_name);
+            EXECUTE FORMAT ('CREATE USER MAPPING FOR %I SERVER %I', role_name, server_internal);
+        EXCEPTION WHEN OTHERS THEN
+            RAISE EXCEPTION 'Could not create server %: %', server, SQLERRM
+                USING HINT = 'Please clean the left over objects';
+        END;
+    END IF;
+
+    -- Add new options
+    FOR row IN SELECT p.key, p.value from lateral json_each_text(final_config->'server') p
+    LOOP
+        IF NOT EXISTS (
+            WITH a AS (
+                SELECT split_part(unnest(srvoptions), '=', 1) AS options FROM pg_foreign_server WHERE srvname=server_internal
+            ) SELECT * from a where options = row.key)
+        THEN
+            EXECUTE FORMAT('ALTER SERVER %I OPTIONS (ADD %I %L)', server_internal, row.key, row.value);
+        ELSE
+            EXECUTE FORMAT('ALTER SERVER %I OPTIONS (SET %I %L)', server_internal, row.key, row.value);
+        END IF;
+    END LOOP;
+
+    -- Update user mapping settings
+    FOR option IN SELECT o.key, o.value from lateral json_each_text(final_config->'user_mapping') o
+    LOOP
+        IF NOT EXISTS (
+            WITH a AS (
+                SELECT split_part(unnest(umoptions), '=', 1) as options from pg_user_mappings WHERE srvname = server_internal AND usename = role_name
+            ) SELECT * from a where options = option.key)
+        THEN
+            EXECUTE FORMAT('ALTER USER MAPPING FOR %I SERVER %I OPTIONS (ADD %I %L)', role_name, server_internal, option.key, option.value);
+        ELSE
+            EXECUTE FORMAT('ALTER USER MAPPING FOR %I SERVER %I OPTIONS (SET %I %L)', role_name, server_internal, option.key, option.value);
+        END IF;
+    END LOOP;
+END
+$$
+LANGUAGE PLPGSQL VOLATILE PARALLEL UNSAFE;
+
+--
+-- Drops a registered server and all the objects associated with it
+-- 
+CREATE OR REPLACE FUNCTION @extschema@.CDB_Federated_Server_Unregister(server TEXT)
```

Comment:
This affects both `CDB_Federated_Server_Unregister` and `CDB_Federated_Table_Unregister` directly and `CDB_Federated_Table_Register` indirectly (since it calls `CDB_Federated_Table_Unregister`), so I'm going to create an issue with this suggestion since adding it involves also adding quite a bit of tests.

Reply:
Alright. Please cc/ relevant people in the new ticket. We can discuss whether/when to implement it in the issue.

Reply:
It's here: https://github.com/CartoDB/cartodb-platform/issues/6034